### ISSUE TITLE
[Tooltip] Do not hide overlay while tooltip is hovered

### ISSUE THREAD

**Issue by xuanshenbo** [angular/components#4942]
#### Bug, feature request, or proposal:
Feature request

#### What is the expected behavior?
An option that allows not close tooltip when mouse hovers over tooltip.
Something like: https://www.w3schools.com/css/tryit.asp?filename=trycss_tooltip

#### What is the current behavior?
The tooltip is immediately shown when the user's mouse hovers over the element. The tooltip immediately hides when the user's mouse leaves the element, it treats the tooltip area no difference with other area outside the element.

I understand that I can delay the hide time, but that is a different story.

#### What are the steps to reproduce?

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Doc site example: https://material.angular.io/components/component/tooltip

#### What is the use-case or motivation for changing an existing behavior?
As a web user, I found that on many websites, tooltips do not close when I mouse hover over them. This allows me to copy the text in the tooltip.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Material 2.0.0-beta.6

#### Is there anything else we should know?
If it doesn't sound like a fair feature, is there a workaround I can use to achieve this?
I tried 
```
md-tooltip-component:hover {
  .mat-tooltip {
    visibility: visible;
  }
}
```
But it did not work

**Comment by andrewseguin** [comment]
You're right that the hide delay doesn't help here since interactions outside of the trigger (e.g. selecting the tooltip text) will hide the tooltip immediately. We'll need to come back to this another time once we figure out if this fits the general tooltip material design.

**Comment by azazlhr** [comment]
@xuanshenbo any solution or workaround?

**Comment by gtamas** [comment]
@andrewseguin 

I have no idea if this fits the material design or not, but the current behavior doesn't meet WCAG requirements either, so it's actually an A11Y issue. This is especially painful, because WCAG 2.x AA is now a must in the EU for gov sites (among others), it's required by law.

I guess the functionality should be modified to meet the the following requirements:
w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html

Any plans for fixing this in the near future?

**Comment by digeomel** [comment]
I just came across this as well. Still not fixed? Any alternatives/work-arounds?

**Comment by manuelmeister** [comment]
@crisbeto this ticket probably needs the a11y label. So that it is not overlooked in the accessibility epic.

A11Y Requirement: https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html

**Comment by crisbeto** [comment]
This isn't relevant for accessibility, because the tooltip content is hidden from screen readers completely. We use `aria-describedby` to convey the message.

**Comment by manuelmeister** [comment]
This is requirement is not relevant to screen readers, but for accessibility. 
The section [content-on-hover-or-focus.html#hoverable](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html#hoverable) describes the requirement. You can see in the examples:

---

<img src="https://www.w3.org/WAI/WCAG21/Understanding/img/button-pointer-on-tooltip.png" alt="Screenshot of a button with a large mouse pointer over it, and a tooltip displayed  below the button which is obscured by the large pointer">
<img src="https://www.w3.org/WAI/WCAG21/Understanding/img/button-pointer-below-tooltip.png" alt="Screenshot of a button with a tooltip below it, and a large mouse pointer at the bottom of the tooltip">			

Figure 3 A button's tooltip is displayed directly below it on mouse hover which can easily be obscured by a large pointer.  The tooltip itself is able to be hovered so the mouse pointer can be moved down to its bottom edge in order to view the tooltip text.

**Comment by aardrian** [comment]
The [_Basic tooltip_](https://material.angular.io/components/tooltip/overview#:~:text=longpresses%20an%20element.-,Basic%20tooltip,-link) in the docs as I see them today definitely has WCAG bugs:
1. It fails WCAG 2.1 SC [1.4.13 Content on Hover or Focus](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=145%2C244%2C249%2C1412#content-on-hover-or-focus), specifically:
  > Where receiving and then removing pointer hover or keyboard focus triggers additional content to become visible and then hidden, the following are true:
  > […]
  > Hoverable: If pointer hover can trigger the additional content, then the pointer can be moved over the additional content without the additional content disappearing;
  > […]

2. The specific example fails WCAG 2.1 SC [2.5.3 Label in Name](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=145%2C244%2C249%2C1412#label-in-name) because the `aria-label` provides the accName for the control, overriding the visible text:
```
<button […] aria-label="Button that displays a tooltip when focused or hovered over" […]>
  <span class="mat-button-wrapper"> Action </span>
  […]
</button>
```

Obviously # 2 is less a concern because authors can "easily" fix that by not following the example in the docs.

The first issue is a genuine bug in the implementation. Which means every site that implements this tooltip component as-is is automatically failing that WCAG SC, and therefore WCAG as a whole. Which has all sorts of compliance and legal ramifications.

The issue was filed prior WCAG 2.1's release with those SCs, but it was added in the August 2017 Working Draft and went live in June 2018. About 3⅔ years ago.

Since this now requires votes to fix this WCAG conformance bug, consider this comment as my vote, though I rather hope the "bug" label (or equivalent) gets added.

**Comment by jelbourn** [comment]
Removing "feature" this this is a real a11y bug

**Comment by jelbourn** [comment]
@crisbeto got any time to look at this?

**Comment by aardrian** [comment]
This bug report is closed, but I don't see it corrected at the URL where this issue was reported.

**Comment by crisbeto** [comment]
The fix was merged an hour ago so it hasn't been released yet.

**Comment by dricks** [comment]
Hi,
this fixes has a big drawback because now tooltips which are displayed over other buttons or links are really annoying.
Perhaps the behavior should have been to cancel the hideTimer if the cursor moves over the tooltip before the timer expires, and start the hide timer again when the cursor leaves the tooltip, without adding any overlay between the target element and the tooltip

**Comment by aardrian** [comment]
> this fixes has a big drawback because now tooltips which are displayed over other buttons or links are really annoying.

This is one reason why tool-tips are often frowned upon by experienced UI folks. They either hide valuable information behind an interaction, or present superfluous information and shouldn't be used.

As a developer, you can generally improve UIs by making them unnecessary and removing them. As a user, you can press `Esc` until those designs catch up.

**Comment by T4YoO** [comment]
Is that under option ? Because we need to keep the previous behavior..

**Comment by aardrian** [comment]
The previous behavior introduced a WCAG violation (per [comment](#issuecomment-1027324047)).

**Comment by crisbeto** [comment]
There's the `disableTooltipInteractivity` config option if you want to disable the new behavior for your app.

**Comment by SadiinsoSnowfall** [comment]
I also encounter the "tooltip are now displayed over other buttons" problem, especially when buttons are presented in mat-table columns. @crisbeto overriding the default `disableTooltipInteractivity` would be a global (as in it would affect the whole app/component) if I'm not mistaken. Would it be possible to add such an option to individual matTooltips ?

**Comment by crisbeto** [comment]
It shouldn't be a problem to provide it on a per tooltip basis, I just figured that this is something that you either want or don't want for the entire app. For what it's worth, you can also provide it at the component level through the `providers` property on `@Component`.

**Comment by realnatuki** [comment]
This is ridiculous; my brother is one of 0.5%  people using custom large cursors giving me the same feedback. I'm building apps for 99.5% of normal people using standard ui setup! This break every layout that we have. I've posted a video example on another PR.
